Kit changes for Bootstrap Commands

piyush-jena opened the pull request:
<!--
Tips:
- Please read CONTRIBUTING.md to understand our process and our requests for PRs.
- Please file an issue before creating a PR so we can discuss the change and confirm it's not already being worked on.
-->

**Issue number:**

Closes #

**Description of changes:**
This PR contains core-kit changes to add `bootstrap-commands`. The relevant settings-sdk changes are in [PR 46](https://github.com/bottlerocket-os/bottlerocket-settings-sdk/pull/46)



**Testing done:**
Test case 1: 2 bootstrap commands
```
[settings.bootstrap-commands."002-test-bootstrap-commands"]
commands = [["apiclient", "set", "motd=helloworld2"], ["apiclient", "set", "motd=helloworld3"]]
mode = "once"
essential = true

[settings.bootstrap-commands."001-test-bootstrap-commands"]
commands = [["apiclient", "set", "motd=helloworld1"]]
mode = "always"
essential = true
```
first boot sets `motd` to `helloworld3`
second boot sets `motd` to `helloworld1`

Test case 2: No bootstrap commands
No boot issues.

Note:
3 files have been modified to build successfully. "sources/api/Cargo.toml", "sources/deny.toml", "sources/Cargo.lock". These are dependent on the settings-sdk changes in [PR 46](https://github.com/bottlerocket-os/bottlerocket-settings-sdk/pull/46). They will be reverted back at the end of code review.

**Terms of contribution:**

By submitting this pull request, I agree that this contribution is dual-licensed under the terms of both the Apache License, version 2.0, and the MIT license.

arnaldo2792 commented on packages/os/bootstrap-commands.service:
```diff
@@ -0,0 +1,18 @@
+
+[Service]
+Type=oneshot
+# Run the bootstrap commands
```
nit: this comment is kind of extra, I think what's happening is self-explanatory.

arnaldo2792 commented on sources/api/bootstrap-commands/Cargo.toml:
```diff
@@ -0,0 +1,37 @@
+toml = "0.8"
+
+[dependencies.bottlerocket-modeled-types]
+git = "https://github.com/piyush-jena/bottlerocket-settings-sdk"
```
I think you can split this dependency update to a separate commit, with title `hack` so that we know it has to be changed after you release a new version of the settings SDK that includes your change.

piyush-jena replied:
https://github.com/bottlerocket-os/bottlerocket-project-website/issues/473

piyush-jena replied:
updated.

arnaldo2792 commented on sources/api/bootstrap-commands/README.md:
```diff
@@ -0,0 +1,28 @@
+# bootstrap-commands
```
You will have to have a "website" task to update the documentation for both bootstrap commands and the `apiclient` actions.

piyush-jena replied:
https://github.com/bottlerocket-os/bottlerocket-project-website/issues/473

arnaldo2792 commented on sources/api/bootstrap-commands/src/main.rs:
```diff
@@ -0,0 +1,345 @@
+through the Bottlerocket API. It introduces a new systemd service that facilitates the execution of
+these commands, expanding the functionality of the existing Bottlerocket API.
+
+Currently, Bottlerocket provides a set of actions (such as refresh-updates and reboot) that can be
```
I wouldn't list `reboot` as an action that you would like to run a bootstrap command. You could just leave `refresh-updates`.

arnaldo2792 replied:
Also, this kind of doesn't match with the example you provide. In here, you say that bootstrap commands allow one to run "actions". However, in your example you run `apiclient set` which isn't in itself an action.

arnaldo2792 commented on sources/api/bootstrap-commands/src/main.rs:
```diff
@@ -0,0 +1,345 @@
+    let mut command = Command::new(bin_path);
+    command.args(args);
+    let output = command
+        .output()
```
Rather than creating new pipes for STDIN/STDOUT, we should use the parent's process pipes with `status` so that the child process inherits the pipes, but the parent process waits for it to complete.

piyush-jena replied:
updated

arnaldo2792 commented on sources/api/bootstrap-commands/src/main.rs:
```diff
@@ -0,0 +1,345 @@
+        .output()
+        .context(error::ExecutionFailureSnafu { command })?;
+
+    trace!("stdout: {}", String::from_utf8_lossy(&output.stdout));
+    trace!("stderr: {}", String::from_utf8_lossy(&output.stderr));
```
If you do what I suggested above, you wouldn't need this.

piyush-jena replied:
Updated.

arnaldo2792 commented on sources/api/bootstrap-commands/src/main.rs:
```diff
@@ -0,0 +1,345 @@
+            return Ok(());
+        }
+
+        // Create a file in /run/bootstrap-commands to indicate an attempt to run this bootstrap-command
+        command(constants::TOUCH_BIN, [temp_path])?;
```
I don't think you need to create a sentry file for bootstrap commands. If they are `once` in the API you assume they haven't run. And the `bootstrap-commands` systemd service won't be re-executed in later stages of the boot (systemd guarantees that). In bootstrap containers, we needed that file because for failed bootstrap containers systemd will attempt to re-run them as it moved to other targets. You don't have that problem.

arnaldo2792 commented on sources/api/bootstrap-commands/src/main.rs:
```diff
@@ -0,0 +1,345 @@
+where
+    S: AsRef<str>,
+{
+    let binding = Vec::new();
```
Why is this called `binding`?

piyush-jena replied:
removed.

arnaldo2792 commented on sources/api/bootstrap-commands/src/main.rs:
```diff
@@ -0,0 +1,345 @@
+            let args = cmd.get_args();
+            match args {
+                Ok(arguments) => {
+                    if let Err(_e) = command(API_CLIENT_PATH, arguments) {
```
Is this how we are enforcing that only `apiclient` commands are executed? If that's the case, remove the `apiclient` from your example, and just use the actual command executed by it.

piyush-jena replied:
Updated.

arnaldo2792 commented on sources/api/bootstrap-commands/src/main.rs:
```diff
@@ -0,0 +1,345 @@
+        info!("Processing Bootstrap command '{}' ... ", name);
+        for cmd in commands.iter() {
+            // this condition is already taken care of
+            let args = cmd.get_args();
```
You can do:

```
let args = cmd.get_args().context(error::BootstrapCommandFailureSnafu { name })?;
```

And avoid the `match` arms, which is also incorrect since the way it is written doesn't actually trigger an error.

piyush-jena replied:
The discussed `get_args` has a default value which makes the function infallible which is why I decided not to use snafu in this case.

arnaldo2792 commented on sources/api/bootstrap-commands/src/main.rs:
```diff
@@ -0,0 +1,345 @@
+
+    // We want to run bootstrap commands in sorted order of their names.
+    for name in bootstrap_commands.keys().sorted() {
+        let bootstrap_command = &bootstrap_commands[name];
```
I wonder if we should turn on the same clippy rule we used in the Bottlerocket Settings SDK, and avoid accessing elements in a `HashMap` like this, even though it is safe since you are getting the keys from the actual hash_map.

piyush-jena replied:
I couldn't find a better way to get key, value where keys are in sorted order.

piyush-jena replied:
Updated HashMap to BTreeMap which stores keys in BST (sorted order). Verified it works with end to end testing.

arnaldo2792 commented on sources/api/bootstrap-commands/src/main.rs:
```diff
@@ -0,0 +1,345 @@
+    // We want to run bootstrap commands in sorted order of their names.
+    for name in bootstrap_commands.keys().sorted() {
+        let bootstrap_command = &bootstrap_commands[name];
+        if let Err(e) = handle_bootstrap_command(name, bootstrap_command) {
```
It is common practice in this repo to use Snafu to provide context for the errors, instead of evaluating if the function call returned an error.

piyush-jena replied:
Updated.

arnaldo2792 commented on packages/os/bootstrap-commands-toml:
```diff
@@ -0,0 +1,12 @@
+{{#if_not_null settings.bootstrap-commands}}
+{{#each settings.bootstrap-commands}}
+[bootstrap-commands."{{@key}}"]
+commands = [{{join_nested_array this.commands}}]
```
Why do we need the helper? Can't we just pass `this.commands`?

piyush-jena replied:
I attempted. It didn't work.

cbgbt replied:
I think this has the capacity to output invalid `toml` -- if the inputs have any characters in `[]"`, then there may be problems.

I believe you can use `'` for raw strings in TOML, and triple-single-quotes for raw strings which allow `'` as well.

Might be a good idea to write a test case in the `bootstrap-commands` binary which checks against input like this.

cbgbt replied:
As an alternative to `join_nested_array`, I wonder if you could kill two birds with one stone and implement it in the template like this:

```
{{#each settings.bootstrap-commands}}
[bootstrap-commands."{{@key}}"]
commands = [
{{#each this.commands}}
    [{{#each this}}'''{{ this }}''',{{/each}}],
{{/each}}
]
{{/each}}
```

It seems to [work on the handlebars playground](https://handlebarsjs.com/playground.html#format=1&currentExample=%7B%22template%22%3A%22commands%20%3D%20%5B%5Cn%7B%7B%23each%20commands%7D%7D%5Cn%20%20%5B%7B%7B%23each%20this%7D%7D'''%7B%7B%20this%20%7D%7D'''%2C%7B%7B%2Feach%7D%7D%5D%2C%5Cn%7B%7B%2Feach%7D%7D%5Cn%5D%5Cn%22%2C%22partials%22%3A%5B%5D%2C%22input%22%3A%22%7B%20commands%3A%20%5B%5B%5C%22apiclient%5C%22%2C%20%5C%22get%5C%22%2C%20%5C%22settings.motd%5C%22%5D%2C%20%5B%5C%22apiclient%5C%22%2C%20%5C%22set%5C%22%2C%20%5C%22settings.motd%3Dhi%5C%22%5D%5D%20%7D%5Cn%22%2C%22output%22%3A%22commands%20%3D%20%5B%5Cn%20%20%5B'''apiclient'''%2C'''get'''%2C'''settings.motd'''%2C%5D%2C%5Cn%20%20%5B'''apiclient'''%2C'''set'''%2C'''settings.motd%26%23x3D%3Bhi'''%2C%5D%2C%5Cn%5D%5Cn%22%2C%22preparationScript%22%3A%22Handlebars.registerHelper('loud'%2C%20function%20(aString)%20%7B%5Cn%20%20%20%20return%20aString.toUpperCase()%5Cn%7D)%5Cn%22%2C%22handlebarsVersion%22%3A%224.7.8%22%7D), other than the unusual URL encoding. Also seems to parse properly per [this site](https://handlebarsjs.com/playground.html#format=1&currentExample=%7B%22template%22%3A%22commands%20%3D%20%5B%5Cn%7B%7B%23each%20commands%7D%7D%5Cn%20%20%5B%7B%7B%23each%20this%7D%7D'''%7B%7B%20this%20%7D%7D'''%2C%7B%7B%2Feach%7D%7D%5D%2C%5Cn%7B%7B%2Feach%7D%7D%5Cn%5D%5Cn%22%2C%22partials%22%3A%5B%5D%2C%22input%22%3A%22%7B%20commands%3A%20%5B%5B%5C%22apiclient%5C%22%2C%20%5C%22get%5C%22%2C%20%5C%22settings.motd%5C%22%5D%2C%20%5B%5C%22apiclient%5C%22%2C%20%5C%22set%5C%22%2C%20%5C%22settings.motd%3Dhi%5C%22%5D%5D%20%7D%5Cn%22%2C%22output%22%3A%22commands%20%3D%20%5B%5Cn%20%20%5B'''apiclient'''%2C'''get'''%2C'''settings.motd'''%2C%5D%2C%5Cn%20%20%5B'''apiclient'''%2C'''set'''%2C'''settings.motd%26%23x3D%3Bhi'''%2C%5D%2C%5Cn%5D%5Cn%22%2C%22preparationScript%22%3A%22Handlebars.registerHelper('loud'%2C%20function%20(aString)%20%7B%5Cn%20%20%20%20return%20aString.toUpperCase()%5Cn%7D)%5Cn%22%2C%22handlebarsVersion%22%3A%224.7.8%22%7D).

piyush-jena replied:
I solved it with triple curly braces for now but I would like to spend some more time on this to think about a better way to do this to tackle corner cases that we discussed.

cbgbt commented on sources/api/bootstrap-commands/src/main.rs:
```diff
@@ -0,0 +1,315 @@
+*/
+
+#[macro_use]
+extern crate log;
```
You are no longer required to use `macro_use` as of Rust 2018 edition.

cbgbt commented on sources/api/bootstrap-commands/src/main.rs:
```diff
@@ -0,0 +1,315 @@
+#[macro_use]
+extern crate log;
+
+//use itertools::Itertools;
```
```suggestion
```

cbgbt commented on sources/api/bootstrap-commands/src/main.rs:
```diff
@@ -0,0 +1,315 @@
+use snafu::{ensure, OptionExt, ResultExt};
+use std::collections::BTreeMap;
+
+use std::env;
```
nit: Seems like the whitespace may not actually be separating dependencies like it was intended.

In general (to my knowledge), Bottlerocket coding style has dependencies existing without whitespace separation and allows `cargo fmt` to sort them alphabetically.

The exception would be when using `pub use` to re-export symbols -- those are often kept separate.

cbgbt commented on sources/api/bootstrap-commands/src/main.rs:
```diff
@@ -0,0 +1,315 @@
+
+        assert_eq!(bootstrap_commands, expected_bootstrap_commands)
+    }
+}
```
Do you mind also testing some negative cases?

e.g. what happens with an empty input? One with an invalid command?

cbgbt commented on sources/api/bootstrap-commands/src/main.rs:
```diff
@@ -0,0 +1,315 @@
+        },
+
+        #[snafu(display("Essential Bootstrap command with name: {} failed.", name))]
+        BootstrapCommandFailure { name: String },
```
It seems counter-intuitive, but with `snafu` it tends to make sense to drop words like "Failure" and "Error" for enum variants, since `snafu` creates the `context` types like `BoostrapCommandSnafu`.

```suggestion
        BootstrapCommand { name: String },
```

`BootstrapCommandFailureSnafu` is a bit redundant 😄

piyush-jena replied:
Updated it to `BootstrapCommandExecutionSnafu`

cbgbt commented on sources/api/bootstrap-commands/src/main.rs:
```diff
@@ -0,0 +1,315 @@
+{
+    let mut command = Command::new(bin_path);
+
+    command
+        .args(args)
+        .status()
+        .context(error::ExecutionFailureSnafu { command })?;
```
I wonder if we should capture the `stdout/stderr` streams somewhere?

Forwarding `stderr` to this process' `stderr` in a way that makes it clear that it has been passed through might be a good start, e.g. something like:

```rust
info!("[command {} stderr] {}", command_id, line);
```

piyush-jena replied:
[Link](https://doc.rust-lang.org/std/process/struct.Command.html#method.status) status() inherits stdout and stderr of the parent.

cbgbt replied:
Oh, I didn't know! Nice.

cbgbt commented on sources/api/bootstrap-commands/src/main.rs:
```diff
@@ -0,0 +1,315 @@
+#[derive(Debug, PartialEq, Serialize, Deserialize)]
+#[serde(rename_all = "kebab-case")]
+struct BootstrapCommand {
+    #[serde(default, skip_serializing_if = "Option::is_none")]
```
Does the `default` specifier have an effect for `Option` types? If you just plan to `unwrap_or_default` these, you could keep the `default` specifier and drop the `Option`. This would tell `serde` to use `Default::default` to choose the value if the input does not have a value present.

Do we ever serialize this? Since `bootstrap-commands` isn't a library and doesn't have cause to serialize `BootstrapCommands`, maybe we should just remove the `Serialize` derive and the serialization-related attributes.

cbgbt commented on sources/api/bootstrap-commands/src/main.rs:
```diff
@@ -0,0 +1,315 @@
+#[derive(Serialize, Deserialize)]
+#[serde(rename_all = "kebab-case")]
+struct BootstrapCommandConfig {
+    bootstrap_commands: Option<BTreeMap<Identifier, BootstrapCommand>>,
```
This could be a good case to use `#[serde(default)]` if you wanted to avoid dealing with an additional `Option` type, since it's semantically equivalent to the empty list.

I don't have a strong preference though, this is just a suggestion.

cbgbt commented on sources/api/bootstrap-commands/src/main.rs:
```diff
@@ -0,0 +1,315 @@
+{
+    let name = name.as_ref();
+    let mode = bootstrap_command.mode.clone().unwrap_or_default();
+    let commands = bootstrap_command.commands.clone().unwrap_or_default();
```
```suggestion
    let mode = bootstrap_command.mode.clone().unwrap_or_default();
    let commands = bootstrap_command.commands.as_deref().unwrap_or_default();
```

If you drop the `Option` from this struct and use `#[serde(default)]`, you could use `as_ref()` instead of `clone()` for the `mode` as well.

cbgbt commented on sources/api/bootstrap-commands/src/main.rs:
```diff
@@ -0,0 +1,315 @@
+        info!("Processing Bootstrap command '{}' ... ", name);
+
+        for command in commands.iter() {
+            // this condition is already taken care of
```
It's not clear to me what this comment is referring to.

cbgbt commented on sources/api/bootstrap-commands/src/main.rs:
```diff
@@ -0,0 +1,315 @@
+    Ok(())
+}
+
+/// Query the API for the currently defined bootstrap commands
```
```suggestion
/// Read our config file for the set of defined bootstrap commands
```

cbgbt commented on sources/api/bootstrap-commands/src/main.rs:
```diff
@@ -0,0 +1,315 @@
+}
+
+/// Query the API for the currently defined bootstrap commands
+fn get_bootstrap_commands<P>(config_path: P) -> Result<BTreeMap<Identifier, BootstrapCommand>>
```
The ordering enforced by `BTreeMap` is important to the correctness of this program, but it's also easy to miss.

Suggestion: have this function return the `BootstrapCommandConfig` struct wholesale. Add a method on it like,

```rust
fn iter(&self) -> impl Iterator<Item = (Identifier, BootstrapCommand)> {
}
```

...and ensure that the docstring mentions that the iterator is sorted lexicographically by ID. This hides the `BTreeMap` implementation from callers while also making it clear what the intent is.

cbgbt commented on sources/api/bootstrap-commands/src/main.rs:
```diff
@@ -0,0 +1,315 @@
+    let bootstrap_commands = get_bootstrap_commands(args.config_path)?;
+
+    // We want to run bootstrap commands in sorted order of their names.
+    for (name, bootstrap_command) in bootstrap_commands.iter() {
```
Just to follow up about the `BTree` implementation -- this comment about sorted order seems almost erroneous in isolation unless I go read the function signature and become aware that the return type is `BTreeMap`.

cbgbt commented on sources/api/schnauzer/src/helpers.rs:
```diff
@@ -975,6 +975,86 @@ pub fn join_array(
+                template: template_name,
+            })?;
+
+    let mut result = String::new();
+
+    for (i, array) in nested_array.iter().enumerate() {
+        if i > 0 {
```
This is quite efficient, plus the tests give me confidence. As a future note, I think a quite "rust-y" way to accomplish this may have been to use iterators + [`Vec::join`](https://doc.rust-lang.org/std/vec/struct.Vec.html#method.join), e.g. something like:


```rust
let nested_array = nested_array_param.as_array().with_context(/* etc */)?;
let nested_array: Vec<String> = nested_array.into_iter()
    .map(|inner_array| {
        let inner_array = inner_array.as_array().with_context(/* etc */)?;
        let inner_array: Vec<String> = inner_array.into_iter()
            .map(|value| format!("\"{}\""), value.as_str().context(/* etc */))
            .collect::<Result<Vec<_>>>()?;
        Ok(inner_array.join(", "))
    })
    .collect::<Result<Vec<_>>>()?;
let result = nested_array.join(", ");

piyush-jena replied:
Replaced join_nested_array with toml_encode as discussed.

cbgbt commented:
>It seems to [work on the handlebars playground](https://handlebarsjs.com/playground.html#format=1&currentExample=%7B%22template%22%3A%22commands%20%3D%20%5B%5Cn%7B%7B%23each%20commands%7D%7D%5Cn%20%20%5B%7B%7B%23each%20this%7D%7D'''%7B%7B%20this%20%7D%7D'''%2C%7B%7B%2Feach%7D%7D%5D%2C%5Cn%7B%7B%2Feach%7D%7D%5Cn%5D%5Cn%22%2C%22partials%22%3A%5B%5D%2C%22input%22%3A%22%7B%20commands%3A%20%5B%5B%5C%22apiclient%5C%22%2C%20%5C%22get%5C%22%2C%20%5C%22settings.motd%5C%22%5D%2C%20%5B%5C%22apiclient%5C%22%2C%20%5C%22set%5C%22%2C%20%5C%22settings.motd%3Dhi%5C%22%5D%5D%20%7D%5Cn%22%2C%22output%22%3A%22commands%20%3D%20%5B%5Cn%20%20%5B'''apiclient'''%2C'''get'''%2C'''settings.motd'''%2C%5D%2C%5Cn%20%20%5B'''apiclient'''%2C'''set'''%2C'''settings.motd%26%23x3D%3Bhi'''%2C%5D%2C%5Cn%5D%5Cn%22%2C%22preparationScript%22%3A%22Handlebars.registerHelper('loud'%2C%20function%20(aString)%20%7B%5Cn%20%20%20%20return%20aString.toUpperCase()%5Cn%7D)%5Cn%22%2C%22handlebarsVersion%22%3A%224.7.8%22%7D), other than the unusual URL encoding. Also seems to parse properly per [this site](https://handlebarsjs.com/playground.html#format=1&currentExample=%7B%22template%22%3A%22commands%20%3D%20%5B%5Cn%7B%7B%23each%20commands%7D%7D%5Cn%20%20%5B%7B%7B%23each%20this%7D%7D'''%7B%7B%20this%20%7D%7D'''%2C%7B%7B%2Feach%7D%7D%5D%2C%5Cn%7B%7B%2Feach%7D%7D%5Cn%5D%5Cn%22%2C%22partials%22%3A%5B%5D%2C%22input%22%3A%22%7B%20commands%3A%20%5B%5B%5C%22apiclient%5C%22%2C%20%5C%22get%5C%22%2C%20%5C%22settings.motd%5C%22%5D%2C%20%5B%5C%22apiclient%5C%22%2C%20%5C%22set%5C%22%2C%20%5C%22settings.motd%3Dhi%5C%22%5D%5D%20%7D%5Cn%22%2C%22output%22%3A%22commands%20%3D%20%5B%5Cn%20%20%5B'''apiclient'''%2C'''get'''%2C'''settings.motd'''%2C%5D%2C%5Cn%20%20%5B'''apiclient'''%2C'''set'''%2C'''settings.motd%26%23x3D%3Bhi'''%2C%5D%2C%5Cn%5D%5Cn%22%2C%22preparationScript%22%3A%22Handlebars.registerHelper('loud'%2C%20function%20(aString)%20%7B%5Cn%20%20%20%20return%20aString.toUpperCase()%5Cn%7D)%5Cn%22%2C%22handlebarsVersion%22%3A%224.7.8%22%7D).


I figured out the encoding bit. Apparently handlebars HTML-encodes data by default with the "double-stache". [If you "triple-stache", it embeds it directly](https://handlebarsjs.com/playground.html#format=1&currentExample=%7B%22template%22%3A%22commands%20%3D%20%5B%5Cn%7B%7B%23each%20commands%7D%7D%5Cn%20%20%5B%7B%7B%23each%20this%7D%7D'''%7B%7B%7B%20this%20%7D%7D%7D'''%2C%7B%7B%2Feach%7D%7D%5D%2C%5Cn%7B%7B%2Feach%7D%7D%5Cn%5D%5Cn%22%2C%22partials%22%3A%5B%5D%2C%22input%22%3A%22%7B%20commands%3A%20%5B%5B%5C%22apiclient%5C%22%2C%20%5C%22get%5C%22%2C%20%5C%22settings.motd%5C%22%5D%2C%20%5B%5C%22apiclient%5C%22%2C%20%5C%22set%5C%22%2C%20%5C%22settings.motd%3Dhi%5C%22%5D%5D%20%7D%5Cn%22%2C%22output%22%3A%22commands%20%3D%20%5B%5Cn%20%20%5B'''apiclient'''%2C'''get'''%2C'''settings.motd'''%2C%5D%2C%5Cn%20%20%5B'''apiclient'''%2C'''set'''%2C'''settings.motd%3Dhi'''%2C%5D%2C%5Cn%5D%5Cn%22%2C%22preparationScript%22%3A%22Handlebars.registerHelper('loud'%2C%20function%20(aString)%20%7B%5Cn%20%20%20%20return%20aString.toUpperCase()%5Cn%7D)%5Cn%22%2C%22handlebarsVersion%22%3A%224.7.8%22%7D).

bcressey commented on packages/os/bootstrap-commands.service:
```diff
@@ -0,0 +1,17 @@
+SyslogIdentifier=bootstrap-commands
+
+[Install]
+WantedBy=preconfigured.target
```
This needs to be `RequiredBy` or else failures will not stop the system from reaching `preconfigured.target`.
```suggestion
[Install]
RequiredBy=preconfigured.target
```

bcressey commented on packages/os/os.spec:
```diff
@@ -89,6 +92,7 @@ Requires: %{_cross_os}sundog
 Requires: %{_cross_os}xfscli
 Requires: %{_cross_os}thar-be-settings
 
+Requires: %{_cross_os}bootstrap-commands
```
nit: this can go in the almost alphabetized list above of unconditional dependencies

bcressey commented on sources/Cargo.toml:
```diff
@@ -3,6 +3,7 @@ resolver = "1"
 members = [
     "api/apiserver",
     "api/apiclient",
+    "api/bootstrap-commands",
```
I'd prefer not to put this under "api"; historically, that was the collection of crates that depended directly on API access, which isn't the case for this program.

(Several of the other crates here need to be moved out, but that cleanup can happen separately.)

bcressey commented on sources/api/bootstrap-commands/src/main.rs:
```diff
@@ -0,0 +1,315 @@
+/*
```
This needs to be a doc comment, or it won't populate the template used to render `README.md`
```suggestion
/*!
```

bcressey commented on packages/os/bootstrap-commands-toml:
```diff
@@ -0,0 +1,12 @@
+[bootstrap-commands."{{@key}}"]
+commands = [{{join_nested_array this.commands}}]
+mode = "{{{this.mode}}}"
+essential = {{this.essential}}
```
`commands`, `mode`, and `essential` are not guaranteed to be set - it's possible that only one is set. You need to guard each individual field with `{{#if_not_null ... }}`, in addition to the top-level one.

bcressey commented on sources/api/bootstrap-commands/src/main.rs:
```diff
@@ -0,0 +1,315 @@
+new feature aims to extend the accessibility of these actions by allowing Bottlerocket consumers
+to run them prior to the bootstrap container execution.
+
+#Example:
```
```suggestion
# Example
```

bcressey commented on sources/api/bootstrap-commands/src/main.rs:
```diff
@@ -0,0 +1,315 @@
+Currently, Bottlerocket provides a set of actions (such as refresh-updates) that can be accessed
+through the apiserver, typically by running the apiclient in the control container. However, this
+new feature aims to extend the accessibility of these actions by allowing Bottlerocket consumers
+to run them prior to the bootstrap container execution.
```
Style-wise, I would mimic the style of the `bootstrap-containers/src/main.rs` - describe what the program does at a high-level, in a way that's useful for a developer working on the project and needs context on what this program is supposed to do.

Save the larger context for the website. This program is part of implementing the feature for Bottlerocket, but it is not the feature.

bcressey commented on sources/api/bootstrap-commands/src/main.rs:
```diff
@@ -0,0 +1,315 @@
+essential = true
+mode = "always"
+
+This would update motd from "Welcome to Bottlerocket!" to "helloworld".
```
nit: avoid referring to the default motd value in case it changes:
```suggestion
This would set `/etc/motd` to "helloworld".
```

bcressey commented on sources/api/bootstrap-commands/src/main.rs:
```diff
@@ -0,0 +1,315 @@
+    let mode = bootstrap_command.mode.clone().unwrap_or_default();
+    let commands = bootstrap_command.commands.clone().unwrap_or_default();
+
+    info!("Running bootstrap command with name: {}", name);
```
I don't think we should emit this message unless we're actually running the command. Otherwise the logs may be misleading if someone just greps for "Running bootstrap command".

bcressey commented on sources/api/bootstrap-commands/src/main.rs:
```diff
@@ -0,0 +1,315 @@
+            run_command("apiclient", ["set", formatted.as_str()])?;
+        }
+
+        info!("Succesfully ran bootstrap command '{}'", name);
```
```suggestion
        info!("Successfully ran bootstrap command '{}'", name);
```

bcressey commented on sources/api/bootstrap-commands/src/main.rs:
```diff
@@ -0,0 +1,315 @@
+    Ok(())
+}
+
+/// Handles running of the api commands in a Bootstrap command
```
nit: this comment doesn't add much value, it just restates the function name

bcressey commented on sources/api/bootstrap-commands/src/main.rs:
```diff
@@ -0,0 +1,315 @@
+            run_command(cmd.as_str(), arg_str)?;
+        }
+
+        if mode == "once" {
+            let formatted = format!("settings.bootstrap-commands.{}.mode=off", name);
+            info!("Turning off bootstrap-command '{}'", name);
+            run_command("apiclient", ["set", formatted.as_str()])?;
+        }
```
The problem with this approach is that if one of the commands is `apiclient reboot`, then we may not finish changing the mode before the reboot kicks in. Same (but more subtly) if the command is `apiclient update apply -c -r` for a one-time update at boot.

To fix this, you can do something like:
1. Check for `/var/lib/bottlerocket/bootstrap-commands@<name>.ran` prior to executing a "once" command. If found, skip the command.
2. Create `/var/lib/bottlerocket/bootstrap-commands@<name>.ran` prior to executing a "once" command. This records that the command has been initiated.
3. For any "off" command, remove the marker file if it exists, as the user might later re-enable it as a "once" command and expect it to run again.

This will satisfy the first part of the contract - making sure a command only runs once.

For the second part, updating the API settings to "off" to reflect that the commands have run, you could maintain a set of bootstrap commands to disable, and disable them all at once prior to returning a value from the overall program execution.

piyush-jena replied:
We opted on solving this with systemd-inhibit.

bcressey commented on sources/api/bootstrap-commands/src/main.rs:
```diff
@@ -0,0 +1,315 @@
+
+    // We want to run bootstrap commands in sorted order of their names.
+    for (name, bootstrap_command) in bootstrap_commands.iter() {
+        let essential = bootstrap_command.essential.unwrap_or(false);
+        let status = handle_bootstrap_command(name, bootstrap_command);
+
+        ensure!(
+            !essential || status.is_ok(),
+            error::BootstrapCommandFailureSnafu {
+                name: name.as_ref()
+            }
+        )
```
This intersects with the "once" behavior in a strange way: if the first set of commands is marked as "essential" and "once", and it fails, then the boot should fail.

But, we have to decide what to do on a subsequent boot:
1. We already ran the command once, so we don't have to run it again, even though it wasn't successful. This exposes later commands in the list to an unexpected state - maybe they expect previous commands to have all been executed successfully, but on the next boot we'll skip over the failed ones.
2. We ran the command already, but it failed, so we need to run it again. This implies "once" means "one successful execution". This is easier for later commands in the list to reason about - they can be sure that previous essential commands were all executed successfully.

I lean toward (2) as it's consistent with how bootstrap containers are handled; we don't disable a "once" container unless it returns success.

Relative to my earlier comment, you can implement this by creating `/var/lib/bottlerocket/bootstrap-commands@<name>.ran` after the command returns success, rather than before.

bcressey replied:
> Relative to my earlier comment, you can implement this by creating /var/lib/bottlerocket/bootstrap-commands@<name>.ran after the command returns success, rather than before.

Ugh, but this re-introduces the problem of "what if the command causes a reboot".

I would look at [systemd-inhibit](https://www.freedesktop.org/software/systemd/man/latest/systemd-inhibit.html) instead - if you wrap the entire `bootstrap-commands` program call with this, it should inhibit any reboots that might be initiated until you exit.

That should avoid any need to deal with marker files and the like.

piyush-jena replied:
Resolved.

bcressey commented on sources/api/bootstrap-commands/src/main.rs:
```diff
@@ -0,0 +1,315 @@
+            source: toml::de::Error,
+        },
+
+        #[snafu(display("Essential Bootstrap command with name: {} failed.", name))]
```
nit: avoid saying "name" in error messages, and just give the quoted identifier:
```suggestion
        #[snafu(display("Essential bootstrap command '{}' failed.", name))]
```

cbgbt commented on packages/os/bootstrap-commands-toml:
```diff
@@ -0,0 +1,22 @@
+[bootstrap-commands."{{@key}}"]
+{{#if_not_null this.commands}}
+commands = [
+{{#each this.commands}}
+    [{{#each this}}'''{{{ this }}}''',{{/each}}],
+{{/each}}
```
I think this will work for most cases, unless a user uses `'''` for some reason. That would be unusual but not inconceivable.

There are two other ideas I had:
* Write a helper which encodes an input `serde_json::Value` as an inline toml value using [the toml crate](https://docs.rs/toml/).
* Embrace HTML escaping, go back to using "double stache", and then have the `bootstrap-commands` binary decode the HTML-escaped input using [the `html-escape` crate](https://crates.io/crates/html-escape)

cbgbt commented on packages/os/bootstrap-commands-toml:
```diff
@@ -0,0 +1,22 @@
+{{#each settings.bootstrap-commands}}
+[bootstrap-commands."{{@key}}"]
+{{#if_not_null this.commands}}
+commands = [
+{{#each this.commands}}
+    {{ toml_encode this }},
+{{/each}}
+]
```
No need to iterate anymore, now that you have `toml_encode`!

```suggestion
commands = {{ toml_encode this.commands }}
```

bcressey commented on packages/os/os.spec:
```diff
@@ -246,6 +250,12 @@ Requires: %{_cross_os}binutils
 %description -n %{_cross_os}driverdog
 %{summary}.
 
+
```
nit: unnecessary extra newline
```suggestion
```

bcressey commented on packages/os/os.spec:
```diff
@@ -640,6 +652,13 @@ install -p -m 0644 %{S:400} %{S:401} %{S:402} %{buildroot}%{_cross_licensedir}
 %{_cross_bindir}/certdog
 %{_cross_templatedir}/certdog-toml
 
+
```
```suggestion
```

bcressey commented on packages/systemd/systemd.spec:
```diff
@@ -14,6 +14,7 @@ Source3: journald.conf
 Source4: issue
 Source5: systemd-journald.conf
 Source6: systemd-sysusers.conf
+Source7: systemd-logind.conf
```
nit: I would prefer to put this in `release` where we've collected other Bottlerocket-specific configuration files for `systemd`.

bcressey commented on packages/systemd/systemd-logind.conf:
```diff
@@ -0,0 +1,2 @@
+[Login]
+InhibitDelayMaxSec=300
```
Can you add a comment on why this is being set to 5 minutes?

bcressey commented on sources/api/schnauzer/src/helpers/stdlib/mod.rs:
```diff
@@ -1130,3 +1130,127 @@ mod test_negate_or_else {
     }
 }
+
+/// `toml_encode` accepts arbitrary input and encodes it as a toml string
```
Can you split the new `toml_encode` helper into a different commit?

bcressey commented on packages/os/bootstrap-commands.service:
```diff
@@ -0,0 +1,18 @@
+# We depend on systemd-logind.service for running systemd-inhibit.
+After=systemd-logind.service settings-applier.service apiserver.service
+Requires=systemd-logind.service settings-applier.service apiserver.service
+Before=preconfigured.target
```
This is a default dependency because it's required by `preconfigured.target`:
```suggestion
```

bcressey commented on sources/bootstrap-commands/src/main.rs:
```diff
@@ -0,0 +1,367 @@
+## Example:
+Given a bootstrap command called `001-test-bootstrap-commands` with the following configuration:
+
+[settings.bootstrap-commands.001-test-bootstrap-commands]
+commands = [[ "apiclient", "set", "motd=helloworld"]]
+essential = true
+mode = "always"
```
Can you apply a code format block to these examples?

bcressey commented on sources/bootstrap-commands/src/main.rs:
```diff
@@ -0,0 +1,367 @@
+This would set `/etc/motd` to "helloworld".
+
+# Additional Information:
+Certain valid apiclient commands that work in a session may fail in bootstrap-commands due to
```
nit: let's code-format mentions of `apiclient`, `bootstrap-commands` and `bootstrap-commands.service` , so it's clear that we're talking about specific programs or units and not something more general.
```suggestion
Certain valid `apiclient` commands that work in a session may fail in `bootstrap-commands` due to
```

bcressey commented on sources/bootstrap-commands/src/main.rs:
```diff
@@ -0,0 +1,367 @@
+
+bootstrap-commands ensures that Bootstrap Commands are executed as defined in the system
+settings. It is called by bootstrap-commands systemd service which runs prior to the
+execution of bootstrap-containers.
```
```suggestion
`bootstrap-commands` ensures that Bootstrap Commands are executed as defined in the system
settings. It is called by `bootstrap-commands.service` which runs prior to the
execution of `bootstrap-containers`.
```

bcressey commented on sources/bootstrap-commands/src/main.rs:
```diff
@@ -0,0 +1,367 @@
+settings. It is called by bootstrap-commands systemd service which runs prior to the
+execution of bootstrap-containers.
+
+Each bootstrap command is a set of commands. The service queries the API for the settings, then
```
"The service queries the API for the settings" is not accurate; the settings are rendered into a config file.

bcressey commented on sources/bootstrap-commands/src/main.rs:
```diff
@@ -0,0 +1,367 @@
+
+This command fails because bootstrap-commands.service which calls this binary is launched prior
+to preconfigured.target while host-containers@.service which is a requirement for running "exec"
+commands are launched after preconfigured.target.
```
This doesn't feel like the right place to put commentary on how the systemd units are configured, which is handled elsewhere. That said, this is also where `README.md` comes from, so I understand why the content is here and maybe there's not a better place for it.

piyush-jena replied:
This addition was to resolve this [comment](https://github.com/bottlerocket-os/bottlerocket-settings-sdk/pull/46#discussion_r1733637534). This restriction seemed like something that the user should be informed which is why I wanted to add it in README.

bcressey commented on sources/bootstrap-commands/src/main.rs:
```diff
@@ -0,0 +1,367 @@
+    let mode = bootstrap_command.mode.as_ref();
+    let commands = &bootstrap_command.commands;
+
+    if mode == "off" {
+        // If mode is 'off', just log this information.
+        info!("Bootstrap command mode for '{}' is 'off'", name);
```
nit: can do an early return to outdent the remaining code:
```suggestion
    if mode == "off" {
        // If mode is 'off', just log this information.
        info!("Bootstrap command mode for '{}' is 'off'", name);
        return  Ok(());
    }
```

bcressey commented on sources/bootstrap-commands/src/main.rs:
```diff
@@ -0,0 +1,367 @@
+        // If mode is 'off', just log this information.
+        info!("Bootstrap command mode for '{}' is 'off'", name);
+    } else {
+        info!("Processing Bootstrap command '{}' ... ", name);
```
```suggestion
        info!("Processing bootstrap command '{}' ... ", name);
```

bcressey commented on sources/bootstrap-commands/src/main.rs:
```diff
@@ -0,0 +1,367 @@
+
+        if mode == "once" {
+            let formatted = format!("settings.bootstrap-commands.{}.mode=off", name);
+            info!("Turning off bootstrap-command '{}'", name);
```
```suggestion
            info!("Turning off bootstrap command '{}'", name);
```

piyush-jena commented:
Forced push addresses the following:
1. Changes to `sources/bootstrap-commands/README.md`
2. nit comments in `sources/bootstrap-commands/main.rs`
3. Removes unnecessary `Before=preconfigured.target` in `packages/os/bootstrap-commands.service`

piyush-jena committed:
schnauzer: add toml_encode helper

piyush-jena committed:
bootstrap-commands: implement bootstrap command execution

piyush-jena commented:
Forced push addresses the following:
1. Split the commits to separate out `schnauzer` and `bootstrap-commands` changes.
2. Moved drop-in config for logind from `systemd.spec` to `release.spec`

cbgbt commented on sources/bootstrap-commands/src/main.rs:
```diff
@@ -0,0 +1,370 @@
+
+impl BootstrapCommandConfig {
+    // Gets an iterator for bootstrap_commands, sorted in lexicographical order of their names.
+    fn iter(self) -> impl Iterator<Item = (Identifier, BootstrapCommand)> {
+        self.bootstrap_commands.into_iter()
+    }
+}
```
I think this is fine, but as a Rust tip, there's the [`IntoIter`](https://doc.rust-lang.org/stable/std/iter/trait.IntoIterator.html) in `std` that you can implement for this behavior, which can make calling other APIs that expect `IntoIter` ergonomic.

cbgbt reviewed (approved):
Nice work! It's probably worth documenting how reboots are handled -- that might be best on the website.

piyush-jena commented:
forced push removes the hack commit to use the official settings-sdk

arnaldo2792 reviewed (approved):
Nice job! :tada:
